src/typescript/frontend/src/middleware.ts

```diff
@@ -21,6 +34,35 @@ export default async function middleware(request: NextRequest) {
   if (MAINTENANCE_MODE && pathname !== ROUTES.maintenance) {
     return NextResponse.redirect(new URL(ROUTES.maintenance, request.url));
   }
+
+  if (pathname.startsWith(ROUTES.api["."])) {
+    const ip = request.ip ?? "127.0.0.1";
+
+    let limit: number;
+    let reset: number;
+    let remaining: number;
+    let success: boolean;
+
+    const ratelimitRes = await ratelimit.limit(ip);
+    limit = ratelimitRes.limit;
+    reset = ratelimitRes.reset;
+    remaining = ratelimitRes.remaining;
+    success = ratelimitRes.success;
+
+    const headers = {
+      "X-RateLimit-Limit": limit.toString(),
+      "X-RateLimit-Remaining": remaining.toString(),
+      "X-RateLimit-Reset": reset.toString(),
+    };
+
+    if (!success) {
+      return new Response("", {
```

Comment:
```suggestion
      return new NextResponse("", {
```

just to keep it consistent